## Упражнения по библиотеке Numpy

In [2]:
import math
import time
import numpy as np
from scipy import spatial

In [3]:
def random_array(fr=-10, t=10, size=10):
    return np.random.randint(fr, t, size=size)

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [4]:
array = random_array()
print(array)
array = np.where(np.logical_and(array > 3, array < 8), array * -1, array)
array

[-4  2  1  5  2  3 -6  2 -9 -7]


array([-4,  2,  1, -5,  2,  3, -6,  2, -9, -7])

**2.** Заменить максимальный элемент случайного массива на 0

In [5]:
array = random_array()
print(array)
array[array == max(array)] = 0
array

[ 4  3 -1  4 -5  6  9 -6  7  5]


array([ 4,  3, -1,  4, -5,  6,  0, -6,  7,  5])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [6]:
def cartesian(arrays):
    arrays = [np.asarray(arr) for arr in arrays]
    shape = map(len, arrays)
    idx = np.indices(shape, dtype=int)
    idx = idx.reshape(len(arrays), -1).T
    for n, arr in enumerate(arrays):
        idx[:, n] = arrays[n][idx[:, n]]
    return idx


cartesian([[1, 2, 3], [4, 5, 6]])

array([[1, 4],
       [1, 5],
       [1, 6],
       [2, 4],
       [2, 5],
       [2, 6],
       [3, 4],
       [3, 5],
       [3, 6]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [7]:
a = random_array(size=(8, 3))
print(a)
b = random_array(size=(2, 2))
print(b)
mask = [np.isin(row, b[0]).any() and np.isin(row, b[1]).any() for row in a]
a[mask]

[[  8  -2  -8]
 [  7   8  -3]
 [-10  -3   5]
 [  9  -8  -2]
 [ -5  -8 -10]
 [  2   3   1]
 [  0  -8  -6]
 [ -6  -2   5]]
[[-4 -5]
 [ 9  5]]


array([], shape=(0, 3), dtype=int64)

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [8]:
matrix = random_array(fr=1, t=3, size=(10, 3))
print(matrix)
to_delete = np.logical_and.reduce(matrix[:, 1:] == matrix[:, :-1], axis=1)
matrix[~to_delete]

[[2 1 1]
 [1 2 1]
 [1 2 1]
 [2 1 2]
 [1 1 2]
 [1 1 1]
 [1 2 2]
 [1 1 1]
 [1 2 2]
 [1 2 1]]


array([[2, 1, 1],
       [1, 2, 1],
       [1, 2, 1],
       [2, 1, 2],
       [1, 1, 2],
       [1, 2, 2],
       [1, 2, 2],
       [1, 2, 1]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [9]:
matrix = random_array(fr=1, t=3, size=(10, 3))
print(matrix)
np.unique(matrix, axis=0)

[[2 1 2]
 [2 2 2]
 [2 1 1]
 [1 2 2]
 [1 1 1]
 [1 1 1]
 [1 1 1]
 [1 1 2]
 [2 2 1]
 [1 2 2]]


array([[1, 1, 1],
       [1, 1, 2],
       [1, 2, 2],
       [2, 1, 1],
       [2, 1, 2],
       [2, 2, 1],
       [2, 2, 2]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [10]:
x = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
prod = 1
for i in range(len(x)):
    for j in range(len(x[i])):
        if i == j and x[i][j] != 0:
            prod *= x[i][j]
prod

3

In [11]:
x = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
diag = x.diagonal()
prod = np.prod(diag[diag != 0])
prod

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [12]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
sorted(x) == sorted(y)

True

In [13]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
np.array_equal(x.sort(), y.sort())

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [14]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
maximum = float('-inf')
for i in range(len(x) - 1):
    if x[i] == 0 and maximum < x[i + 1]:
        maximum = x[i + 1]
maximum

5

In [15]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
is_zero = x == 0
x[1:][is_zero[:-1]].max()

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [16]:
x = [2, 2, 2, 3, 3, 3, 5]
nums = [x[0]]
counts = [0]
for i in range(len(x)):
    if nums[-1] != x[i]:
        nums.append(x[i])
        counts.append(1)
    else:
        counts[-1] += 1

(nums, counts)

([2, 3, 5], [3, 3, 1])

In [17]:
x = np.array([2, 2, 2, 3, 3, 3, 5, ])
np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [18]:
random_array(0, 10, (3, 3))

array([[0, 2, 7],
       [7, 6, 4],
       [2, 9, 7]])

In [19]:
x = random_array(0, 10, (3, 2)).tolist()
print(x)
y = random_array(0, 10, (3, 2)).tolist()
print(y)
start = time.time()
result = []
dim = len(x[0])
for j in range(len(y)):
    result.append([])
    for i in range(len(x)):
        s = 0
        for k in range(dim):
            s += (x[i][k] - y[j][k]) ** 2
        result[j].append(round(math.sqrt(s), 2))
end = time.time()
(result, end - start)

[[9, 1], [3, 7], [3, 0]]
[[2, 9], [2, 1], [7, 4]]


([[10.63, 2.24, 9.06], [7.0, 6.08, 1.41], [3.61, 5.0, 5.66]],
 0.00016307830810546875)

In [20]:
x = random_array(0, 10, (3, 2))
print(x)
y = random_array(0, 10, (3, 2))
print(y)
start = time.time()
x1 = np.repeat(x, len(y), axis=0).reshape(len(x), -1, 2)
y1 = np.repeat(y, len(x), axis=0).reshape(len(y), -1, 2).transpose(1, 0, 2)
result = np.sqrt(np.sum((x1 - y1) ** 2, axis=2)).round(2)
end = time.time()
(result, end - start)

[[9 9]
 [3 0]
 [9 5]]
[[3 7]
 [1 9]
 [0 6]]


(array([[6.32, 8.  , 9.49],
        [7.  , 9.22, 6.71],
        [6.32, 8.94, 9.06]]),
 0.00019478797912597656)

In [21]:
x = np.random.randint(0, 10, (3, 2))
print(x)
y = np.random.randint(0, 10, (3, 2))
print(y)
start = time.time()
result = spatial.distance.cdist(x, y)
end = time.time()
(result, end - start)

[[5 3]
 [2 5]
 [8 4]]
[[5 7]
 [1 4]
 [6 2]]


(array([[4.        , 4.12310563, 1.41421356],
        [3.60555128, 1.41421356, 5.        ],
        [4.24264069, 7.        , 2.82842712]]),
 0.00010251998901367188)

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [22]:
calorie_stats = np.loadtxt("../data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [23]:
my_calories = 60.
average_calories = np.mean(calorie_stats)
delta_calories = np.round(abs(my_calories - average_calories))
delta_calories

47.0

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [24]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [25]:
median_calories = np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [26]:
for i in range(101):
    percentile_value = np.percentile(calorie_stats, i)
    if percentile_value > 60:
        nth_percentile = i
        break
nth_percentile

4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [27]:
more_calories = np.round(100*len(calorie_stats[calorie_stats > 60])/len(calorie_stats))
more_calories

96.0

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [28]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Калорий меньше, чем у 96% конкурентов

Стандартное отклонение в 19.4 каллории говорит об умеренном разбросе данных, следовательно данные равномерны

Батончики CrunchieMunchies опережают большинство других конкурентов по уровню каллорийности, при грамотном маркетинге они обречены на успех